src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRefreshQueue.cs

```diff
@@ -0,0 +1,204 @@
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    internal class SemanticTokensRefreshQueue :
+        IOnInitialized,
+        ILspService,
+        IDisposable
+    {
+        /// <summary>
+        /// Lock over the mutable state that follows.
+        /// </summary>
+        private readonly object _gate = new();
+
+        /// <summary>
+        /// Mapping from project id to the workqueue for producing the corresponding compilation for it on the OOP server.
+        /// </summary>
+        private readonly Dictionary<ProjectId, CompilationAvailableEventSource> _projectIdToEventSource = new();
+
+        /// <summary>
+        /// Mapping from project id to the project-cone-checksum for it we were at when the project for it had its
+        /// compilation produced on the oop server.
+        /// </summary>
+        private readonly Dictionary<ProjectId, Checksum> _projectIdToLastComputedChecksum = new();
+
+        private readonly LspWorkspaceManager _lspWorkspaceManager;
+        private readonly IClientLanguageServerManager _notificationManager;
+
+        private readonly IAsynchronousOperationListener _asyncListener;
+        private readonly CancellationTokenSource _disposalTokenSource;
+
+        /// <summary>
+        /// Debouncing queue so that we don't attempt to issue a semantic tokens refresh notification too often.
+        /// 
+        /// Null when the client does not support sending refresh notifications.
+        /// </summary>
+        private AsyncBatchingWorkQueue<Uri?>? _semanticTokenRefreshQueue;
+
+        private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
+
+        public SemanticTokensRefreshQueue(
+            IAsynchronousOperationListenerProvider asynchronousOperationListenerProvider,
+            LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+            LspWorkspaceManager lspWorkspaceManager,
+            IClientLanguageServerManager notificationManager)
+        {
+            _asyncListener = asynchronousOperationListenerProvider.GetListener(FeatureAttribute.Classification);
+
+            _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
+            _disposalTokenSource = new();
+
+            _lspWorkspaceManager = lspWorkspaceManager;
+            _notificationManager = notificationManager;
+        }
+
+        public Task OnInitializedAsync(ClientCapabilities clientCapabilities, CancellationToken _)
+        {
+            if (_semanticTokenRefreshQueue is null && clientCapabilities.Workspace?.SemanticTokens?.RefreshSupport is true)
```

Comment:
@dibarbet here's my fix for the race issue you were talking about. Let me know if I missed anything. Also curious about your thoughts on the `IOnInitialized` interface.

Reply:
seems absolutely OK to me!